### ISSUE TITLE
Sealed TextFieldDisplayDriver in 2.0 makes upgrading difficult

### ISSUE THREAD

**Issue by brunoAltinet** [OrchardCMS/OrchardCore#16886]
### Describe the bug

I've reached out about this on Discord, but I feel this should be considered a bug and maybe better to discuss it here.

I would inherit/extend a TextFieldDisplayDriver, where i would use both 
.IsNew property and .TypePartDefinition properties to modify the resulting shape. Now that TextFieldDisplayDriver is sealed, i have to move that logic somewhere else. 
I considered moving it to IShapeTableProvider but i don't have either of the former properties. I would hate to copy/paste all of the TextFieldDriver logic just to alter some displays.
Doing it inside the template would prolly have me in the same problems.

### Orchard Core version

2.0

### To Reproduce
Try customizing behaviour of TextFieldDisplayDriver, usually done by inheritance.

### Expected behavior
There should be a good way to extend Content Field behaviour. Either add the same context to Shape events or some other way. No idea.  I support having these classes sealed as long as the same thing can be done in a different way.

### Logs and screenshots
n/a

**Comment by Piedone** [comment]
Have you tried adding a second, custom driver (a part/field can have any number of drivers)?

**Comment by brunoAltinet** [comment]
I would have to have that driver affect the same shape without actually creating a new shape. What used to happen, when i didn't deregister the default one, is it would create 2 inputs. I am not sure how to have 2 drivers work on the same shape?

**Comment by brunoAltinet** [comment]
From what i understand you, 2 drivers would create 2 shapes, i would ignore/hide 1 shape and show the other?
But that would mean i'd have to copy all of the logic from original driver (since i can't inherit it) and do my own customization.
I've looked through the code, I don't see how i can have same shape result processed by different drivers.

Shape Events (ondisplaying for instance) look like a better way of doing this, but as i said, no way to access the stuff i need, even though it could be provided (that context is still around when those events are getting fired).

**Comment by MikeAlhayek** [comment]
@brunoAltinet it would be nice to share your code so we can understand your logic on why you really need to inherit from a display driver. Maybe you are looking for a custom editor?

**Comment by brunoAltinet** [comment]
Sure: 
It's a bit convoluted but there are a couple of use cases:
1. make a field read only after initial creation depending on user privilege
2. customize labels for the same field depending on parent content item type

see here:
https://github.com/altibiz/ugp-web/blob/273098622442cb5b5e2f3d693ee21aa7937e06c3/Members/Base/PartFieldSettings/PartTextFieldDriver.cs#L34

and here:

https://github.com/altibiz/ugp-web/blob/273098622442cb5b5e2f3d693ee21aa7937e06c3/Members/Base/DriverService.cs#L38

**Comment by MikeAlhayek** [comment]
@brunoAltinet can't you just add a new editor that wraps that logic instead of having to inherit and register yet another service?

**Comment by Piedone** [comment]
These are both better done from `IShapeTableProvider`:

- `AdminAttribute.IsApplied()` check: you can hide (place) shapes with any custom logic, see an example [here](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/0684bb884fdc86c510422cc7bc5f4b786899e9bc/Lombiq.TrainingDemo/Services/ShapeHidingShapeTableProvider.cs#L43).
- Customize label: in `OnDisplaying` you can pass new properties to the shape, including such a value, and then use a custom shape override to use that label. I don't recommend setting the corresponding `ContentPartFieldSettings` since that's a global config for the field, but that you can do there too (as well as from an `IContentFieldHandler`).

**Comment by brunoAltinet** [comment]
> These are both better done from `IShapeTableProvider`:
> 
> - `AdminAttribute.IsApplied()` check: you can hide (place) shapes with any custom logic, see an example [here](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/0684bb884fdc86c510422cc7bc5f4b786899e9bc/Lombiq.TrainingDemo/Services/ShapeHidingShapeTableProvider.cs#L43).
> - Customize label: in `OnDisplaying` you can pass new properties to the shape, including such a value, and then use a custom shape override to use that label. I don't recommend setting the corresponding `ContentPartFieldSettings` since that's a global config for the field, but that you can do there too (as well as from an `IContentFieldHandler`).

How do I know if the shape is for a new contentitem? And second, I'd need access to parent part settings, is that possible?
That was pretty straightforward from driver

**Comment by brunoAltinet** [comment]
> @brunoAltinet can't you just add a new editor that wraps that logic instead of having to inherit and register yet another service?
 Tried it but wrapping doesnt work, Prefix is not sert on a wrapped component.

**Comment by Piedone** [comment]
> How do I know if the shape is for a new contentitem? And second, I'd need access to parent part settings, is that possible? That was pretty straightforward from driver

The examples you've linked to don't utilize `IsNew`. But you can also access that from an `IContentDisplayHandler.BuildEditorAsync()` implementation.

You have access to the whole content item, including the field, from content field shapes. If you inspect the context with the debugger in the even, you'll see what's available.

**Comment by brunoAltinet** [comment]
> > How do I know if the shape is for a new contentitem? And second, I'd need access to parent part settings, is that possible? That was pretty straightforward from driver
> 
> The examples you've linked to don't utilize `IsNew`. But you can also access that from an `IContentDisplayHandler.BuildEditorAsync()` implementation.
> 
> You have access to the whole content item, including the field, from content field shapes. If you inspect the context with the debugger in the even, you'll see what's available.

The Editor context is handed over further down, and IsNew is used there. See here: https://github.com/altibiz/ugp-web/blob/273098622442cb5b5e2f3d693ee21aa7937e06c3/Members/Persons/PersonPart.cs#L51

I am not sure what you mean access that from an BuildEditorAsync implementation.
There is no isNew boolean on content field shapes, and there's no way of knowing (from what i can see) if the contentItem is a new one other than that flag on the driver, which is not propaggated further.

To iterate, i need access to parent part definition and if the content item is new at the same time to determine how the field is shown.
Example of request: Member email field should be shown as read-only to regular users after initial input. 
Overriden driver had everything i needed. Now i need to either completely copy the code from that driver or do something else. Shape events does not have isNew keyword.

I suggest handing over isNew to shape events somehow, or even the whole context. I'll make an example PR.

**Comment by Piedone** [comment]
If you implement `IContentDisplayHandler.BuildEditorAsync()`, which is an extension point (much like content part handlers) for use cases like yours, then you can access `IsNew` as well, since it receives `BuildEditorContext`.

**Comment by Piedone** [comment]
So, I don't think any change should be needed for your use case, let alone in general in the magnitude of https://github.com/OrchardCMS/OrchardCore/pull/16898. So, please tell why you think it's still necessary.

**Comment by brunoAltinet** [comment]
I'm still working on this, intermittently, will report back tomorrow on how I find this solution.

________________________________

**Comment by Piedone** [comment]
Thank you!

Please update your PR with:

> As for PR, I would omit the TypePartDefinition but would keep the IsNew flag for better developer experience if anyone wants to extend content field functionality.

> I'd also base those Field Viewmodels on Shape if you think that's ok.

I'm really not sure about that. Do we have similar cases? In any case, this should be a separate PR.

**Comment by sebastienros** [comment]
> I would omit the TypePartDefinition but would keep the IsNew flag

Why? Is this not useful anywhere? Or is it because it can be retrieved in other way (it's not contextual to the curent view).

**Comment by Piedone** [comment]
As in, omit TypePartDefinition in the PR.

**Comment by brunoAltinet** [comment]
> > I would omit the TypePartDefinition but would keep the IsNew flag
> 
> Why? Is this not useful anywhere? Or is it because it can be retrieved in other way (it's not contextual to the curent view).

It can be retrieved from PartFieldDefintion.ContentTypePartDefinition

**Comment by sebastienros** [comment]
Not sure this was mentioned, if you can't alter an existing driver, you can always unregister it and add your own instead. It being sealed would not be a problem for that.
